stdlib/stringio.rb

```diff
@@ -1,116 +1,979 @@
+    @closed == :write || closed?
+  end
+
+  def each(sep = $/, limit = nil, chomp: false)
+    # Calls the block with each remaining line read from the stream; returns self.
+    unless sep.nil?
+      if limit.nil? && sep.is_a?(::Numeric)
+        limit = sep
+        sep = $/
       end
+    end
+    raise(ArgumentError, "if limit is given, it must be greater than 0") if limit && limit < 1
+
+    return enum_for(:each, sep, limit, chomp: chomp) unless block_given?
+
+    global_last_line = $_ # this does not seem right, but makes specs pass, maybe it is right
+
+    got_text = false
+    while (line = gets(sep, limit, chomp: chomp))
+      next if !got_text && sep == '' && line =~ /\A\n+\z/
+      got_text = true
+      yield(line)
+    end
+
+    $_ = global_last_line
+    self
+  end
+
+  def each_byte
+    # Calls the given block with each byte (0..255) in the stream; returns self.
+    return enum_for :each_byte unless block_given?
+    while (byte = getbyte)
+      yield(byte)
+    end
+    self
+  end
+
+  def each_char(&block)
+    # With a block given, calls the block with each remaining character in the stream
+    return enum_for :each_char unless block_given?
+    while (chr = getc)
+      yield(chr)
+    end
+    self
+  end
+
+  def each_codepoint(&block)
+    # With a block given, calls the block with each remaining codepoint in the stream
+    return enum_for(:each_codepoint) unless block_given?
+    while (chr = getc)
+      yield(chr.ord)
+    end
+    self
+  end
+
+  alias each_line each
+
+  def eof
+    # Returns true if the stream is positioned at its end, false otherwise;
+    `check_readable(self)`
+    @eof
+  end
+
+  alias eof? eof
+
+  def external_encoding
+    # Returns the Encoding object that represents the encoding of the stream,
+    # or nil if the stream is in write mode and no encoding is specified.
+    @ext_enc
+  end
+
+  def fcntl
+    # Raises NotImplementedError.
+    raise ::NotImplementedError
+  end
+
+  def fileno
+    # Returns nil. Just for compatibility to IO.
+    nil
+  end
+
+  def flush
+    # Returns an object itself. Just for compatibility to IO.
+    self
+  end
+
+  def fsync
+    # Returns 0. Just for compatibility to IO.
+    0
+  end
+
+  def getbyte
+    # Reads and returns the next 8-bit byte from the stream; see Byte IO.
+    `check_readable(self)`
+    s = @buffer.size
+    return nil if @eof || @pos >= s
+    byte = `self.buffer.data_view.getUint8(self.pos)`
+    @pos += 1
+    @eof = true if @pos >= s
+    byte
+  end
+
+  def getc
+    # Reads and returns the next character from the stream; see Character IO.
+    `check_readable(self)`
+    s = @buffer.size
+    return nil if @eof || @pos >= s
+    c = @buffer.get_raw_string(@pos, `Math.min(16, s - self.pos)`, @ext_enc) # account for multi byte chars
+    c = c[0]
+    @pos += c.bytesize
+    @eof = true if @pos >= s
+    c
+  end
+
+  def gets(sep = $/, limit = nil, chomp: false)
```

Comment:
Ruby StringIO has IO::generic_readable and IO::generic_writable as ancestors - which seem like to be an internal thing.

Still we may want to simulate it somewhat, so that certain functions are not duplicated.